api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ ****************************************************************************
+\page page_scatter_gather_emulation Emulating x86 Scatter and Gather Instructions
+
+\tableofcontents
+
+# Background
+
+The x86 gather and scatter instructions were introduced in the AVX2 and AVX512 instruction set extensions. They allow loading or storing a subset of elements in a vector from/to multiple non-contiguous addresses.
+
+AVX2 has only gather instructions, no scatter instructions, whereas AVX512 has both. AVX2 is limited to 256-bit length vectors, whereas AVX512 has 512-bit support. Both support masking of individual memory accesses, using either a special mask register in AVX512 or another vector in AVX2.
+
+Examples of these instructions are (in DR’s IR):
+```
+vpgatherdd %rax(,%ymm11,4)[4byte] %ymm13 -> %ymm12 %ymm13
+```
+
+Above is an AVX2 gather instruction that reads 32-bit doublewords into the 256-bit `ymm12` vector from addresses generated by adding the base address in `rax` to the corresponding index elements in `ymm11`, conditionally based on the masks in `ymm13`. The mask is set to zero on completion, unless the instruction causes an exception.
+
+```
+vpscatterdd {%k1} %xmm10 -> %rax(,%xmm11,4)[4byte] %k1
+```
+
+Above is an AVX512 scatter instruction that writes 32-bit doublewords from the 128-bit `xmm10` vector to addresses generated by adding the base address in `rax` to the corresponding index element in `xmm11`, conditionally based on the mask reg `k1`. The mask register is set to zero on completion, unless the instruction causes an exception.
+
+
+# Problem Statement
+
+Scatter and gather instructions pose a challenge to DynamoRIO clients that observe memory addresses (like drcachesim, the DR client that collects memory address and control flow traces). They are complex to handle because:
```

Comment:
I would also give broad examples of tool classes people are familiar with: taint tracking tools; address tracing tools

Reply:
Done.